web/modules/weather_data/src/Service/WeatherDataService.php

```diff
@@ -491,6 +506,98 @@ public function getPlaceFromGrid($wfo, $x, $y, $self = false)
+     * of a given observation station to a specified reference
+     * geometry.
+     *
+     * For now the reference geometry is the polygon of
+     * a WFO cell.
+     *
+     */
+    public function getObsDistanceInfo($sourceGeometry, $obs, $index = 0)
+    {
+        // We need to find the closest point in the sourceGeometry
+        // to the observation point
+        $distance = INF;
+        $closest = null;
+        foreach ($sourceGeometry as $sourcePoint) {
+            $lonDiff = $obs->geometry->coordinates[0] - $sourcePoint->lon;
+            $latDiff = $obs->geometry->coordinates[1] - $sourcePoint->lat;
+            $hyp = hypot($lonDiff, $latDiff);
+            if ($hyp < $distance) {
+                $distance = $hyp;
+                $closest = $sourcePoint;
+            }
+        }
+
+        $obsText =
+            "POINT(" .
+            $obs->geometry->coordinates[0] .
+            " " .
+            $obs->geometry->coordinates[1] .
+            ")";
+        $sourcePointText = "POINT(" . $closest->lon . " " . $closest->lat . ")";
+
+        $sourceGeomPoints = array_map(function ($point) {
+            return $point->lon . " " . $point->lat;
+        }, $sourceGeometry);
+        $sourceGeomPoints = implode(", ", $sourceGeomPoints);
+        $sourceGeomText = "POLYGON((" . $sourceGeomPoints . "))";
+
+        $sql =
+            "SELECT ST_DISTANCE_SPHERE(" .
+            "ST_GEOMFROMTEXT('" .
+            $obsText .
+            "'), " .
+            "ST_GEOMFROMTEXT('" .
+            $sourcePointText .
+            "')) as distance, " .
+            "ST_WITHIN(ST_GEOMFROMTEXT('" .
+            $obsText .
+            "'), " .
+            "ST_GEOMFROMTEXT('" .
+            $sourceGeomText .
+            "')) as within;";
+
+        $result = $this->database->query($sql)->fetch();
+        $distanceInfo = [
+            "distance" => (float) $result->distance,
+            "withinGridCell" => !!(int) $result->within,
+            "obsPoint" => [
+                "lon" => $obs->geometry->coordinates[0],
+                "lat" => $obs->geometry->coordinates[1],
+            ],
+            "obsStation" => $obs->properties->station,
+            "stationIndex" => $index,
+        ];
+
+        return $distanceInfo;
+    }
+
+    /**
+     * Logs a serialized version of an obsDistanceInfo array
+     *
+     * (See getObsDistanceInfo() for how this array is produced)
+     */
+    public function logObservationDistanceInfo($obsDistanceInfo)
+    {
+        $promise = $this->newRelic->sendMetric(
+            "wx.observation",
+            $obsDistanceInfo["distance"],
+            [
+                "withinGridCell" => $obsDistanceInfo["withinGridCell"],
+                "stationIndex" => $obsDistanceInfo["stationIndex"],
+                "obsStation" => $obsDistanceInfo["obsStation"],
+                "distance" => $obsDistanceInfo["distance"],
+            ],
+        );
+
+        $promise->wait();
```

Comment:
I saw the comment about this, but I don't understand what purpose it serves. My naive take is it should be safe to fire-and-forget and it can finish whenever it finishes, but I'm not confident at all because of the nature of PHP. 😬

Reply:
Yes, fire and forget would be the way to go. Unfortunately after hacking at it for a while I discovered that Guzzle [doesn't support that](https://stackoverflow.com/a/44774990). I'm thinking we will need to set up a queue that polls on a different thread and put all of our future async requests into it